src/ng/rootScope.js

```diff
@@ -353,11 +353,13 @@ function $RootScopeProvider() {
        *    - `newVal` contains the current value of the `watchExpression`
        *    - `oldVal` contains the previous value of the `watchExpression`
        *    - `scope` refers to the current scope
-       * @param {boolean=} objectEquality Compare for object equality using {@link angular.equals} instead of
-       *     comparing for reference equality.
+       * @param {boolean=|function()} objectEquality Compare for object equality using {@link angular.equals} instead of
+       *     comparing for reference equality. If a function is passed, it will be used as a replacement for {@link angular.equals}.
+       * @param {function()} copier Replacement function for {@link angular.copy} to save value into oldVal for next
+            change event.
        * @returns {function()} Returns a deregistration function for this listener.
        */
-      $watch: function(watchExp, listener, objectEquality) {
+      $watch: function(watchExp, listener, objectEquality, copier) {
```

Comment:
Aha, makes sense !
<sub>Or we can force them tick a checkbox that says "I have read and understood the implications" :grin:</sub>

It would be useful to add more docs about it and how to use the feature properly (maybe an example too). I guess something tbd before merging.

I generally like the idea.

Any benchpress experts around ? Would be useful to make sure that performance isn't affected when you don't use tge feature (e.g. core stuff) and that it gets improved (when used correctly) for user stuff.